Changes from Forked Repo

christie94h opened the pull request:
As discussed, here are the two lines I changed when the document was forked. This allows the main UI to run smoothly.

Commit:
Redefined dummy_data to be equal to dummy_data1.

xlai reviewed (changes requested):
I couldn't get the UI running after the implementation and here is I tried to reproduce it:
```R
setwd('./add_path') # Change the path to the root directory of the project
shiny::runApp(appDir='./app/modules/trial_design/')
```
Which gives me the following error

```R
Listening on http://127.0.0.1:4303
Browsing http://127.0.0.1:4303
Warning in gzfile(file, "rb") :
  cannot open compressed file 'app/data/dummy/dummy_data1.RData', probable reason 'No such file or directory'
Warning: Error in gzfile: cannot open the connection
  63: gzfile
  62: readRDS
   1: shiny::runApp
```

I assume this has to do with the path. Could you include a minimal example demonstrating that it is working as intended?

christie94h commented on app/modules/trial_design/ui.R:
```diff
@@ -6,7 +6,8 @@ library(ggplot2)
 #DUMMY DATA MANIPULATION
 
-dummy_data <- readRDS("app/modules/trial_design/dummy_data1.RData")
+dummy_data <- dummy_data <- readRDS("app/data/dummy/dummy_data1.RData") # CHANGE THIS 
+# I have used dummay_data1 for now - this needs to be changed.
```
It seems that this might be a deeper issue than I originally thought. 

I was running the app using line 177 of the `server.R` file, `shinyApp(ui, server_all)`, which works after previously running both the server and ui separately. 

There seems to be more issues with the code than I thought (loading the data using the `here` package suggests that the file might be corrupted), so I'll start from scratch tomorrow. Is using the code,

`setwd('./add_path') # Change the path to the root directory of the project
shiny::runApp(appDir='./app/modules/trial_design/')`,

as you did above, the standard way of running an app like this?

xlai replied:
No - not necessarily the 'standard' way but my guess is VS Code does this under the hood. Plus this is just temporary as our end goal is to bring everything together under one UI.

Commit:
Changed last line of generate_dummy_data to allow combatability with the here package and ran it to generate the file dummy_data5.RData.

Commit:
Added the here package to the ui.R file and changed the calling of dummy_data5.RData to use the here package.

christie94h commented:
I think I've worked out what the bigger issue is. 

The previous student has coded the main UI in such a way that `sever.R` and `ui.R` aren't the typical server and ui files R expects to see when running a shiny app. Instead, they define several functions, including two called  `server_all` and `ui`. Because `sever.R` and `ui.R` aren't defined in the usual way, the command `shiny::runApp(appDir='./app/modules/trial_design/')` won't run the main UI.

I believe this is done intentionally to link up with the questionnaire like this:
- When the app is opened (using the `shiny::runApp()` command) the questionnaire is shown 
- Once the questionnaire is completed (or if the user wants to go straight to the main UI), the user clicks a button to access the main UI. 
- When this button is pressed, R runs the `ui.R` file, the `server.R` file and the code `shinyApp(ui, server_all)` in that order, which opens a functional main UI.

Regardless of the bigger issue, I've fixed the `dummy_data` issue using the `here` package, along with updating a link between the `generate_dummy_data` function and the `ui.R` file, which you should be able to see now. 

Hopefully what I've explained above makes sense, but do let me know if not and I'll try to fix the issues that I'm forgetting about.

xlai commented on app/data/dummy/generate_dummy_data.r:
```diff
@@ -123,5 +123,5 @@ dummy_data <- generate_dummy_data(
 print(dummy_data)
 head(dummy_data)
-saveRDS(dummy_data, 'dummy_data4.RData')
+save(dummy_data, file = "app/modules/trial_design/dummy_data5.RData", version = 2) # Saving to a file. Using version=2 for compatibility with R 4.0.0 and later
```
i think for consistency this should be saved under app/data/dummy

christie94h replied:
Should be changed to fit that now.

xlai reviewed (changes requested):
just a minor tweak and we are then good to go.

Commit:
Moved dummy_data5.R back to the dummy folder, change the generate_dummy_data function to generate dummy data into the dummy folder and repeated with a new generated dummy data set, dummy_data6.R.

xlai reviewed (changes requested):
the solution I have come up with is to create an global.R file to ensure variables (and subsequently data) that are needed for both ui and server function are within the scope of the project. 

This is just a temporary fix as long as the two UIs are not connected; if the dataset is handled reactively in the server, we wouldn't be in this situation.

Here is my proposed changes to add global.R (which will be removed after the merging of two UIs)

```r
here::i_am('app/modules/trial_design/global.R')

library(shiny)
library(DT)
library(ggplot2)
library(here)

# Data loading
data_env <- new.env()
load(here('app', 'data', 'dummy', 'dummy_data6.RData'), envir = data_env)
dummy_data <- data_env$dummy_data 

# Data processing
dummy_data_trial <- dummy_data$trial
n_doses <- dummy_data_trial[dummy_data_trial$q_variable == "n_doses", "value"]
dummy_data_method <- dummy_data$method
dummy_data_ranking <- dummy_data$ranking
ranking <- c(dummy_data_ranking$method)
ranking <- c("crm", "tpt", "other") # Your test ranking

# Functions
prettify_ranking <- function(ranking_argument) {
  pretty_ranking <- ranking_argument
  pretty_ranking[pretty_ranking == "crm"] <- "CRM"
  pretty_ranking[pretty_ranking == "tpt"] <- "3+3"
  pretty_ranking[pretty_ranking == "other"] <- "Other design (TEST)"
  return(pretty_ranking)
}

parse_params <- function(params_str) {
  params <- strsplit(params_str, ";")[[1]]
  param_list <- lapply(params, function(p) strsplit(p, "=")[[1]])
  names(param_list) <- sapply(param_list, '[', 1)
  sapply(param_list, '[', 2)
}

# Global variables
pretty_ranking <- prettify_ranking(ranking)
questions <- dummy_data
column_names <- sprintf("d(%d)", 1:n_doses)
```

This way, the `server.R` file should be very clean and is left with only the definition of server function (name doesn't matter). Remember to clean up `ui.R` file, taking away those that are already covered in `global.R`

This would then work with any calling, for example `runApp('./app/modules/trial_design/')` .

xlai commented:
@christie94h I've revised the comments above and tested the solution. Let me know when this is completed.

Going forward, we will need to rethink how we want to structure the directories.
https://shiny.posit.co/r/articles/build/app-formats/

Commit:
Added a global.R file and adjusted server.R and ui.R to accomodate.

christie94h commented:
I've added the `global.R` file and changed the server and ui accordingly, but the command `shiny::runApp(appDir='./app/modules/trial_design/')` still doesn't work for me. There isn't a problem with the dummy data anymore, but it no doesn't recognise the ui. This is the error that appears:

```r
Listening on http://127.0.0.1:4865
Browsing http://127.0.0.1:4865
Warning: Error in eval: object 'ui' not found
  43: force
  42: uiHttpHandler
  41: shinyApp
   1: shiny::runApp
Error in eval(exprs, envir) : object 'ui' not found
``` 

Did I miss something when I added the `global.R` file, or did I miss a change that should be made to the `server.R` and/or `ui.R` files?

xlai commented:
> I've added the `global.R` file and changed the server and ui accordingly, but the command `shiny::runApp(appDir='./app/modules/trial_design/')` still doesn't work for me. There isn't a problem with the dummy data anymore, but it no doesn't recognise the ui. This is the error that appears:
> 
> ```r
> Listening on http://127.0.0.1:4865
> Browsing http://127.0.0.1:4865
> Warning: Error in eval: object 'ui' not found
>   43: force
>   42: uiHttpHandler
>   41: shinyApp
>    1: shiny::runApp
> Error in eval(exprs, envir) : object 'ui' not found
> ```
> 
> Did I miss something when I added the `global.R` file, or did I miss a change that should be made to the `server.R` and/or `ui.R` files?

i should have said that those that were put in global.R needs to be removed from ui/server. It seems like it's working now after our last discussion. @christie94h let me know when this is pushed to remote and ready for review

Commit:
Changing global.R to include parts of the server and redefining the server and ui to include comments from previous students.

christie94h commented:
Everything we discussed yesterday has been implemented to this fork, and the code should hopefully run the app using `shiny::runApp()`. Do let me know if not. @xlai I am requesting a final review now.

xlai commented:
A slight issue with this PR - as it is coming from the forked repo and the main is not up-to-date, some files related to `renv` hasn't been pulled into your branch. As a result of that, there is a bunch of files becoming untracked rather than being ignored by `.gitignore`. Let's fix this together.

xlai reviewed (changes requested):
Branch needs to be rebased onto the most up-to-date main before merging

xlai reviewed (approved):
All good. I've put the solution of resolving upstream and origin discrepancy below in case you need to revisit this some point in the future

```
# Add upstream remote if not already added
git remote add upstream <original-repo-url>

# Fetch latest changes from upstream
git fetch upstream

# Switch to your feature branch
git checkout your-feature-branch

# Rebase your commits on top of the latest main
git rebase upstream/main

# Force push to update your PR
git push --force-with-lease origin your-feature-branch
```
